File: TideGraph.java
// TideGraph.java
import java.util.*;

public class TideGraph {
    private static final double AMPLITUDE = 1.5;
    private static final double MEAN_LEVEL = 0.5;
    private static final double PERIOD = 12.42;

    private static double tideLevel(double hours) {
        double rad = (hours / PERIOD) * 2 * Math.PI;
        return MEAN_LEVEL + AMPLITUDE * Math.sin(rad - 1.2);
    }

    private static double[][] generateData() {
        int points = 24 * 6; // every 10 minutes
        double[][] data = new double[points][2];
        for (int i = 0; i < points; i++) {
            double h = i * 10 / 60.0;
            data[i][0] = h;
            data[i][1] = tideLevel(h);
        }
        return data;
    }

    private static List<double[]> findExtrema(double[][] data) {
        List<double[]> highs = new ArrayList<>();
        List<double[]> lows = new ArrayList<>();
        for (int i = 1; i < data.length - 1; i++) {
            if (data[i][1] > data[i-1][1] && data[i][1] > data[i+1][1]) {
                highs.add(data[i]);
            } else if (data[i][1] < data[i-1][1] && data[i][1] < data[i+1][1]) {
                lows.add(data[i]);
            }
        }
        List<double[]> result = new ArrayList<>();
        result.addAll(highs);
        result.addAll(lows);
        return result;
    }

    private static String drawGraph(double[][] data) {
        double minLvl = Double.MAX_VALUE, maxLvl = Double.MIN_VALUE;
        for (double[] d : data) {
            if (d[1] < minLvl) minLvl = d[1];
            if (d[1] > maxLvl) maxLvl = d[1];
        }
        minLvl = Math.floor(minLvl - 0.2);
        maxLvl = Math.ceil(maxLvl + 0.2);

        int height = 20;
        double scale = (maxLvl - minLvl) / height;

        int cols = 50;
        int step = data.length / cols;
        int[] rows = new int[cols];
        double[] times = new double[cols];
        double[] levels = new double[cols];
        for (int i = 0; i < cols; i++) {
            int idx = i * step;
            times[i] = data[idx][0];
            levels[i] = data[idx][1];
            int row = (int)Math.round((maxLvl - levels[i]) / scale);
            if (row < 0) row = 0;
            if (row >= height) row = height - 1;
            rows[i] = row;
        }

        StringBuilder sb = new StringBuilder();
        sb.append("🌊 Tide Graph (24‑hour forecast)\n\n");
        sb.append("Level (m)\n");

        for (int r = 0; r < height; r++) {
            double lvl = maxLvl - r * scale;
            sb.append(String.format("%5.1f |", lvl));
            for (int c = 0; c < cols; c++) {
                sb.append(rows[c] == r ? "*" : " ");
            }
            sb.append("\n");
        }

        sb.append("      +").append("-".repeat(cols)).append("\n");

        // Time labels
        StringBuilder timeLine = new StringBuilder("       ");
        for (int i = 0; i <= 24; i += 3) {
            int pos = (int)((i / 24.0) * cols);
            String label = String.format("%02d:00", i);
            while (timeLine.length() < pos) timeLine.append(" ");
            timeLine.append(label);
        }
        sb.append(timeLine).append("\n\n");

        // Extrema
        List<double[]> highs = new ArrayList<>(), lows = new ArrayList<>();
        for (int i = 1; i < data.length - 1; i++) {
            if (data[i][1] > data[i-1][1] && data[i][1] > data[i+1][1]) highs.add(data[i]);
            else if (data[i][1] < data[i-1][1] && data[i][1] < data[i+1][1]) lows.add(data[i]);
        }
        if (!highs.isEmpty()) {
            sb.append("High tides: ");
            for (int i = 0; i < Math.min(2, highs.size()); i++) {
                if (i > 0) sb.append(", ");
                double h = highs.get(i)[0];
                int hour = (int)h;
                int min = (int)Math.round((h - hour) * 60);
                sb.append(String.format("%02d:%02d (%.1fm)", hour, min, highs.get(i)[1]));
            }
            sb.append("\n");
        }
        if (!lows.isEmpty()) {
            sb.append("Low tides:  ");
            for (int i = 0; i < Math.min(2, lows.size()); i++) {
                if (i > 0) sb.append(", ");
                double h = lows.get(i)[0];
                int hour = (int)h;
                int min = (int)Math.round((h - hour) * 60);
                sb.append(String.format("%02d:%02d (%.1fm)", hour, min, lows.get(i)[1]));
            }
            sb.append("\n");
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        double[][] data = generateData();
        System.out.print(drawGraph(data));
    }
}

File: tide_graph.cpp
// tide_graph.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <string>

const double AMPLITUDE = 1.5;
const double MEAN_LEVEL = 0.5;
const double PERIOD = 12.42;

double tideLevel(double hours) {
    double rad = (hours / PERIOD) * 2 * M_PI;
    return MEAN_LEVEL + AMPLITUDE * sin(rad - 1.2);
}

std::vector<std::pair<double, double>> generateData() {
    std::vector<std::pair<double, double>> data;
    for (int minutes = 0; minutes < 24*60; minutes += 10) {
        double h = minutes / 60.0;
        data.push_back({h, tideLevel(h)});
    }
    return data;
}

std::string drawGraph(const std::vector<std::pair<double, double>>& data) {
    double minLvl = 1e9, maxLvl = -1e9;
    for (auto& p : data) {
        if (p.second < minLvl) minLvl = p.second;
        if (p.second > maxLvl) maxLvl = p.second;
    }
    minLvl = floor(minLvl - 0.2);
    maxLvl = ceil(maxLvl + 0.2);

    int height = 20;
    double scale = (maxLvl - minLvl) / height;

    int cols = 50;
    int step = data.size() / cols;
    std::vector<int> rows(cols);
    std::vector<double> times(cols), levels(cols);
    for (int i = 0; i < cols; i++) {
        int idx = i * step;
        times[i] = data[idx].first;
        levels[i] = data[idx].second;
        int row = round((maxLvl - levels[i]) / scale);
        if (row < 0) row = 0;
        if (row >= height) row = height - 1;
        rows[i] = row;
    }

    std::string result;
    result += "🌊 Tide Graph (24‑hour forecast)\n\n";
    result += "Level (m)\n";

    for (int r = 0; r < height; r++) {
        double lvl = maxLvl - r * scale;
        char buffer[20];
        snprintf(buffer, sizeof(buffer), "%5.1f |", lvl);
        result += buffer;
        for (int c = 0; c < cols; c++) {
            result += (rows[c] == r) ? "*" : " ";
        }
        result += "\n";
    }

    result += "      +";
    result += std::string(cols, '-');
    result += "\n";

    std::string timeLine = "       ";
    for (int i = 0; i <= 24; i += 3) {
        int pos = (int)((i / 24.0) * cols);
        char label[6];
        snprintf(label, sizeof(label), "%02d:00", i);
        while ((int)timeLine.size() < pos) timeLine += " ";
        timeLine += label;
    }
    result += timeLine + "\n\n";

    // Find extrema
    std::vector<std::pair<double, double>> highs, lows;
    for (size_t i = 1; i < data.size() - 1; i++) {
        if (data[i].second > data[i-1].second && data[i].second > data[i+1].second)
            highs.push_back(data[i]);
        else if (data[i].second < data[i-1].second && data[i].second < data[i+1].second)
            lows.push_back(data[i]);
    }

    if (!highs.empty()) {
        result += "High tides: ";
        for (int i = 0; i < (int)std::min(size_t(2), highs.size()); i++) {
            if (i > 0) result += ", ";
            int hour = (int)highs[i].first;
            int min = round((highs[i].first - hour) * 60);
            char buf[20];
            snprintf(buf, sizeof(buf), "%02d:%02d (%.1fm)", hour, min, highs[i].second);
            result += buf;
        }
        result += "\n";
    }
    if (!lows.empty()) {
        result += "Low tides:  ";
        for (int i = 0; i < (int)std::min(size_t(2), lows.size()); i++) {
            if (i > 0) result += ", ";
            int hour = (int)lows[i].first;
            int min = round((lows[i].first - hour) * 60);
            char buf[20];
            snprintf(buf, sizeof(buf), "%02d:%02d (%.1fm)", hour, min, lows[i].second);
            result += buf;
        }
        result += "\n";
    }

    return result;
}

int main() {
    auto data = generateData();
    std::cout << drawGraph(data);
    return 0;
}
